daemon/policy.go

```diff
@@ -302,6 +308,86 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (newRev uint
 	return newRev, nil
 }
 
+type policyAddQueueOptions struct {
+	rules   policyAPI.Rules
+	opts    *AddOptions
+	context *policyQueueContext
+	reFetch bool
+}
+
+type policyQueueContext struct {
+	policyRevision uint64
+	err            error
+	done           chan bool
+	totalTime      spanstat.SpanStat
+	queueTime      spanstat.SpanStat
+}
+
+// policyAddToQueue adds the request to policyAdd in the queue and return a
+// context that will be updated when the done channel is closed.
+func (d *Daemon) policyAddToQueue(rules policyAPI.Rules, opts *AddOptions, reFetch bool) *policyQueueContext {
+	result := &policyQueueContext{
+		policyRevision: 0,
+		err:            nil,
+		done:           make(chan bool),
+	}
+	result.totalTime.Start()
+	result.queueTime.Start()
+
+	d.policyAddQueueRequests <- policyAddQueueOptions{
+		rules:   rules,
+		opts:    opts,
+		reFetch: reFetch,
+		context: result,
+	}
+
+	return result
+}
+
+// PolicyQueueRequestInitilize initialize the daemon request channel and start
+// in backgrond the queue worker
+func (d *Daemon) PolicyQueueRequestInitilize(ctx context.Context) {
+	if d.policyAddQueueRequests == nil {
+		d.policyAddQueueRequests = make(chan policyAddQueueOptions, POLICY_REQUEST_CHANNEL_SIZE)
+	}
+	go d.policyQueueWorker(ctx)
+}
+
+// policyQueueWorker is a queue worker that taking care of PolicyAdd for
+// DNSProxy and Kubernetes watcher to avoid multiples races and lock all policy
+// repo.
+// In case that the channel event has a refetch set to true, the rule will be
+// retrieved from policyRepo to make sure that the newest version is installed
+// correctly and old rules do not overwritte the new ones.
+func (d *Daemon) policyQueueWorker(ctx context.Context) {
+	log.Info("Initialize Policy Queue Worker")
+	for {
+		select {
+		case policyRequest, _ := <-d.policyAddQueueRequests:
+			policyRequest.context.queueTime.End(true)
+			var rules []*policyAPI.Rule
+			rules = policyRequest.rules
+			if policyRequest.reFetch {
+				newRules := policyAPI.Rules{}
+				d.policy.Mutex.Lock()
+				for _, sourceRule := range policyRequest.rules {
+					repoRules := d.policy.SearchRLocked(sourceRule.Labels)
+					newRules = append(newRules, repoRules.DeepCopy()...)
+				}
+				d.policy.Mutex.Unlock()
+				rules, _ = d.dnsRuleGen.GenerateRulesFromSources(newRules)
+			}
+			rev, err := d.PolicyAdd(rules, policyRequest.opts)
+			policyRequest.context.policyRevision = rev
+			policyRequest.context.err = err
+			close(policyRequest.context.done)
+			policyRequest.context.totalTime.End(err != nil)
```

Comment:
Move this to above closing the channel or the spanstat may not be marked as complete by the time the blocked caller resumes